kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,608 @@
+4. If U &lt;= 50%, return U+50 as the score
+5. If 50% &lt; U &lt;= 100%, return 100 - U
+6. If U > 100%, return 0
+
+For example, let’s say we have three nodes X, Y, and Z, with four cores each and utilization 1, 2, and 3 cores respectively. For simplicity, let’s assume our pod to be scheduled has 0 cores CPU requests and overhead.
+
+Utilization of each node:
+
+
+```
+Ux → (1/4)*100 = 25
+Uy → (2/4)*100 = 50
+Uz → (3/4)*100 = 75
+```
+
+
+The score of each node :
+
+
+```
+X → 50 + 25 =  75
+Y → 50 + 50 = 100
+Z → 75 - 50 = 25
+```
+
+
+In the algorithm above, 50% is the target utilization rate we want to achieve on all nodes. We can be less aggressive by reducing it to 40% so that it has much lesser chances of going over 50% during spikes or unexpected loads. 
+
+In the 2nd step of the algorithm, one variant uses the current pod's total CPU limits instead of requests, to have a stronger upper bound of expected utilization.
+
+The 50% threshold value for utilisation will be made configurable via plugin argument.
+
+**Algorithm Analysis**
+
+<img src="images/bfbp-graph.png" alt="bfbp-graph" width="600" height="400"/>
+
+
+The above is a plot of the piecewise function outlined in the algorithm. The key observations here are manifold:
+
+
+
+1. As the utilization goes from 0 to 50%, we pack pods on those nodes by favoring them.
+2. The nodes are penalized linearly as the utilization goes beyond 50%, to spread the pod amongst those "hot" nodes.
+3. The positive slope begins from 50 and not from 0 because the range of score is from 0-100, and we want to maximize our score output for nodes we wish to favor so that the score is more substantial and other plugins do not affect it much.
+4. There is a break in the graph with a high drop due to the "penalty" we have.
+
+#### Spreading with BestFitBinPack
+
+
+To meet the 2nd constraint in the problem statement, we utilize the existing PodTopologySpread (PTS) plugin provided by K8s. 
+This is to avoid baking in another spreading algorithm within BFBP when there is an existing in-tree plugin and conflict our 4th goal.  
+It was proved with experiments that using a dynamic weight-based approach (changing weights of PTS and BFBP in every scheduling cycle) is not a good idea, neither it is supported by Kubernetes. 
+However, if we choose a static large weight for PTS, say 100, we can meet our goals. 
+The problem with this approach is that we will end up downplaying the scores of other important default in-tree plugins like Image Locality, Interpod Affinity, etc. with their default weights of 1. 
+Also, with more plugins we plan to add in the future, managing weights will not be simple. So for BFBP to work well with PTS without modifying weights, 
+the following extended algorithm is proposed that is called when scheduling replicas of a service, replica-set, replication controller, etc.
+
+
+**Algorithm**
+
+
+
+1. For each node, calculate S = (1000\*PTS + PTS\*BFBP)
```

Comment:
are you proposing to do this inside your plugin?

Reply:
Yes.

Reply:
Why? This should be up to users

Reply:
but users can already configure that using the framework

Reply:
I got your point, let me add an option to turn off spreading in docs and use static weights instead as an alternative.